ppd/ppd-test.c

```diff
@@ -0,0 +1,4036 @@
+/*
+ * PPD test program for CUPS.
+ *
+ * Copyright © 2007-2018 by Apple Inc.
+ * Copyright © 1997-2007 by Easy Software Products, all rights reserved.
+ *
+ * Licensed under Apache License v2.0.  See the file "LICENSE" for more
+ * information.
+ *
+ * PostScript is a trademark of Adobe Systems, Inc.
+ */
+
+/*
+ * Include necessary headers...
+ */
+
+#include <cups/cups-private.h>
```

Comment:
Do not use any "private" interfaces from CUPS, they are no stable APIs and only valid for code inside the CUPS of the version you have taken the `cupstestppd.c` from. Please remove this line and do not use any functions whose name starts with `_cups`. Please check these functions and check whether libppd (or libcupsfilters) contains them (or at least similar ones). I have overtaken several of them into libppd (but then renamed them to start with `ppd` or `_ppd`).

Reply:
_cupsLangPuts and _cupsLangPrintf is in langprint.c in cups, an alternative of which is not present in libppd. Should I migrate langprint.c to libppd as well?

Reply:
These functions are to set the user's UI language and to output the translations of given strings into the selected UI language. This way cupstestppd outputs its report in the user language if CUPS has appropriate translations.

In a first approach one could remove the `_cupsSetLocale()` call and replace the other functions by `fputs()` and `fprintf()` and the report will always be in English (the strings in the source code which are supplied to the functions).

Actually you should replace the calls by calls of the log function (which you supply to `ppdTest()` in its parameter list). You will also not have translations with that, but by simply replacing the log function with a more sophisticated one one could easily add translation support later.

But generally, I think translations are not that important here. PPD files are deprecated and will go away in the future, especially when the New Architecture is finally in place. So we should no put too much work into libppd and also, the former `cupstestppd`, provided via your port into libppd will mainly be used by retro-fitting Printer Applications to check the integrity of PPD files or by developers/maintainers of such Printer Applications to handle bug reports and support requests. Developers and maintainers usually understand English and translated logs are even counter-productive for them.